lib/std/start.zig

```diff
@@ -603,14 +603,29 @@ pub fn callMain() u8 {
 }
 
 pub fn call_wWinMain() std.os.windows.INT {
+    const peb = std.os.windows.peb();
     const MAIN_HINSTANCE = @typeInfo(@TypeOf(root.wWinMain)).Fn.params[0].type.?;
     const hInstance = @as(MAIN_HINSTANCE, @ptrCast(std.os.windows.kernel32.GetModuleHandleW(null).?));
     const lpCmdLine = std.os.windows.kernel32.GetCommandLineW();
 
-    // There's no (documented) way to get the nCmdShow parameter, so we're
-    // using this fairly standard default.
-    const nCmdShow = 5;
+    // This makes Zig mostly match the nCmdShow behavior of a C program with a WinMain symbol:
+    // - With STARTF_USESHOWWINDOW set in STARTUPINFO.dwFlags of the CreateProcess call:
+    //   - Compiled with subsystem:console -> nCmdShow is always SW_SHOWDEFAULT
+    //   - Compiled with subsystem:windows -> nCmdShow is STARTUPINFO.wShowWindow from
+    //     the parent CreateProcess call
+    // - With STARTF_USESHOWWINDOW unset:
+    //   - nCmdShow is always SW_SHOWDEFAULT
+    //
+    // TODO: Figure out how to check the subsystem and use it to match the above
+    //       behavior, or intentionally avoid matching the above behavior with
+    //       regards to the subsystem.
+    const SW_SHOWDEFAULT = 10;
+    const STARTF_USESHOWWINDOW = 1;
+    var nCmdShow: std.os.windows.ULONG = SW_SHOWDEFAULT;
+    if (peb.ProcessParameters.dwFlags & STARTF_USESHOWWINDOW != 0) {
+        nCmdShow = peb.ProcessParameters.dwShowWindow;
+    }
 
     // second parameter hPrevInstance, MSDN: "This parameter is always NULL"
-    return root.wWinMain(hInstance, null, lpCmdLine, nCmdShow);
+    return root.wWinMain(hInstance, null, lpCmdLine, @intCast(nCmdShow));
```

Comment:
This `@intCast` is meant to allow for defining `wWinMain` using different types for `nCmdShow`. [The docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-winmain) have `nCmdShow` use the C `int` type, but `ProcessParameters.dwShowWindow` uses the `ULONG` type. This `@intCast` allows for both `c_int` and `u32` to be used when defining a `wWinMain` function on the Zig side.

Reply:
Instead, you should change the type of `nCmdShow` above to be the minimum integer type in which the value fits. If the user picks an integer type that is smaller than that, it will then correctly be a compile error. For example, you could pick `u31` which will coerce into both `i32` and `u32`, but will be a compile error if the user uses `u8`.

Also, please avoid using `var` unless necessary due to there being a loop.

Reply:
The [max recognized value is currently `11`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow). And there's multiple different types that it uses in various APIs:

- `u16` in [STARTUPINFO](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfow) (how the 'show window' value is set during CreateProcess)
- `c_int` in [ShowWindows](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow) (what `nCmdShow` is meant to be used with)
- `u32` in `PEB.ProcessParameters.dwShowWindow` (what Windows populates for the process)

so maybe `u16` would make the most sense since `STARTUPINFO` is the bottleneck?

---

Just so I understand better, could you expand a bit on the avoiding `var` thing?